internal/pkg/EAR/ear_state.go

```diff
@@ -0,0 +1,390 @@
+// Copyright 2021 Google Inc.  All Rights Reserved.
+//
+// This defines a domain for storing sets of equivalent abstract references
+// (which is useful for may-alias relationships). It uses union-find to maintain
+// these partitions. It uses two maps: one from a reference to its parent and
+// one from a partition representative to partition-level information
+// (particularly, field mappings).
+//
+package EAR
+
+import (
+	"log"
+	"sort"
+	"strings"
+)
+
+// ParentMap maps an abstract reference to its representative (i.e.
+// the parent in the union-find tree).
+//
+// (TODO): the maps are not concurrency safe. Use sync.RWMutex for concurrency usage.
+type ParentMap map[AbsReference]AbsReference
+
+// AbsPartitionInfo maintains information at the abstract partition level (which
+// consists of a set of abstract references). It maintains the size of the
+// partition (used for weighted join in union-find) and a AbsReferenceFieldMap,
+// which is a mapping from values F (instance fields, array contents etc) to
+// abstract references R.
+type AbsPartitionInfo struct {
+	numMembers uint                 // number of members
+	fieldRefs  AbsReferenceFieldMap // field mapping
+}
+
+// PartitionInfoMap maps an abstract reference to its partition.
+type PartitionInfoMap map[AbsReference]AbsPartitionInfo
+
+// AbsState is Equivalent Abstract References (EAR) state.
+type AbsState struct {
+	// Map from an abstract ref (a partition representative) to information about
+	// this partition.
+	partitions PartitionInfoMap
+	// Map from an abstract ref to its parent abstract ref.
+	parents ParentMap
+
+	// Set to true after all mutation operations have completed: only lookups will
+	// be performed from here on.
+	finalized bool
```

Comment:
How badly do we need this? Typestate in the contract adds a lot of complexity.

This seems to create a totally new thing when finalized is called. Maybe return object of the intended type as the result of Finalize().

Reply:
"finalized" is to control the mutability of the state: when we finish building the state, we shall not mutate it anymore, and the allowed operations are reads. 
While we can surely create another object to hold these fields such as "parents" and "partitions", it turns out that the same fields are needed. So it is OK to reuse the state.
Either way works for me. If Go allows us to define the state to be "const", then it works better.